src/bos/common/clients/api_client.py

```diff
@@ -0,0 +1,62 @@
+#
+# MIT License
+#
+# (C) Copyright 2021-2024 Hewlett Packard Enterprise Development LP
+#
+# Permission is hereby granted, free of charge, to any person obtaining a
+# copy of this software and associated documentation files (the "Software"),
+# to deal in the Software without restriction, including without limitation
+# the rights to use, copy, modify, merge, publish, distribute, sublicense,
+# and/or sell copies of the Software, and to permit persons to whom the
+# Software is furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+#
+from abc import ABC
+from typing import Type, TypeVar
+
+from bos.common.clients.endpoints import BaseGenericEndpoint
+from bos.common.utils import RetrySessionManager
+
+ClientEndpoint = TypeVar('ClientEndpoint', bound=BaseGenericEndpoint)
+
+
+class APIClient(RetrySessionManager, ABC):
+    """
+    As a subclass of RetrySessionManager, this class can be used as a context manager,
+    and will have a requests session available as self.requests_session
+
+    This context manager is used to provide API endpoints, via subclassing.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._endpoint_values: dict[Type[ClientEndpoint], ClientEndpoint] = {}
```

Comment:
So, the dictionary keys are the Classes themselves? Why did you choose that? Are these keys unique?

Reply:
I mainly did it for ease of type hinting. It made it easier to specify the type signature for the `get_endpoint` method.

I'm not sure I understand the uniqueness question. Classes in Python are unique -- not instances of them, of course, but the classes themselves.

Reply:
No, it can be more than one. For example, the PCS client class has 2 endpoints associated with it -- one for power transitions and one for power status. Look at the `client.py` file in the [PCS PR](https://github.com/Cray-HPE/bos/pull/400/files)

Reply:
Okay, I can see in your example that there are two endpoints in PCS, but they are of different classes. I just didn't want to see two endpoints with the same class where one overwrote the other. 
I was thinking you might have MyGenericEndpoint with EndpointInstance1 and EndpointInstance2 and the second would overwrite the first. But it's looking like they'd all be unique endpoints. I don't think I was clear on what your endpoint classes would look like down the line, and so I didn't realize each endpoint class would be unique and only refer to a single endpoint.